src/Numerical/Spectral/Linearize.cpp

```diff
@@ -0,0 +1,69 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include <array>
+
+#include "Numerical/Spectral/Linearize.hpp"
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/Mesh.hpp"
+#include "DataStructures/StripeIterator.hpp"
+#include "Numerical/Spectral/LegendreGaussLobatto.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/MakeArray.hpp"
+
+template <size_t Dim>
+DataVector linearize(const DataVector& u, const Mesh<Dim>& mesh) {
+  auto u_linearized = make_array<Dim>(DataVector(mesh.product(), 0.0));
+
+  {
+    const Matrix& F = Basis::lgl::linear_filter_matrix(mesh[0]);
+    for (StripeIterator s(mesh, 0); s; ++s) {
+      dgemv_('N', mesh[0], mesh[0], 1., F.data(), mesh[0],
+             u.data() + s.offset(),  // NOLINT
+             s.stride(), 0.0,
+             gsl::at(u_linearized, 0).data() + s.offset(),  // NOLINT
+             s.stride());
+    }
+  }
+
+  for (size_t d = 1; d < Dim; ++d) {
+    const Matrix& F = Basis::lgl::linear_filter_matrix(mesh[d]);
+    for (StripeIterator s(mesh, d); s; ++s) {
+      dgemv_('N', mesh[d], mesh[d], 1., F.data(), mesh[d],
+             gsl::at(u_linearized, d - 1).data() + s.offset(),  // NOLINT
```

Comment:
Since this and the above only differ here in this `gsl::at`, maybe it would be good to use a ternary: `0 == d ? u.data() : gsl::at(u_linearized, d - 1)`? I would put parentheses around the ternary to make sure operator precedence is correct. This will reduce we have to deal with.